File: Languages/go/low-level-design-lld/design-patterns/composite.go
package main

import "fmt"

// Composite Pattern: Compose objects into tree structures to represent
// part-whole hierarchies. Lets clients treat individual objects and
// compositions uniformly.
// Example: File system (files and directories).

// Component interface
type FileSystemNode interface {
	Name() string
	Size() int64
	Display(indent string)
}

// Leaf: File
type File struct {
	name string
	size int64
}

func NewFile(name string, size int64) *File {
	return &File{name: name, size: size}
}

func (f *File) Name() string { return f.name }
func (f *File) Size() int64  { return f.size }
func (f *File) Display(indent string) {
	fmt.Printf("%s📄 %s (%d bytes)\n", indent, f.name, f.size)
}

// Composite: Directory
type Directory struct {
	name     string
	children []FileSystemNode
}

func NewDirectory(name string) *Directory {
	return &Directory{name: name}
}

func (d *Directory) Name() string { return d.name }

func (d *Directory) Size() int64 {
	var total int64
	for _, child := range d.children {
		total += child.Size()
	}
	return total
}

func (d *Directory) Display(indent string) {
	fmt.Printf("%s📁 %s/ (%d bytes)\n", indent, d.name, d.Size())
	for _, child := range d.children {
		child.Display(indent + "  ")
	}
}

func (d *Directory) Add(node FileSystemNode) {
	d.children = append(d.children, node)
}

func (d *Directory) Remove(name string) {
	for i, child := range d.children {
		if child.Name() == name {
			d.children = append(d.children[:i], d.children[i+1:]...)
			return
		}
	}
}

func main() {
	// Build a file tree
	root := NewDirectory("project")

	src := NewDirectory("src")
	src.Add(NewFile("main.go", 1500))
	src.Add(NewFile("handler.go", 2200))
	src.Add(NewFile("model.go", 800))

	tests := NewDirectory("tests")
	tests.Add(NewFile("main_test.go", 900))
	tests.Add(NewFile("handler_test.go", 1100))

	root.Add(src)
	root.Add(tests)
	root.Add(NewFile("go.mod", 120))
	root.Add(NewFile("README.md", 450))

	// Display the tree
	root.Display("")

	// Verify sizes roll up correctly
	expectedSize := int64(1500 + 2200 + 800 + 900 + 1100 + 120 + 450)
	if root.Size() == expectedSize {
		fmt.Printf("\nPASS: total size = %d bytes\n", root.Size())
	} else {
		panic(fmt.Sprintf("FAIL: expected %d, got %d", expectedSize, root.Size()))
	}

	// Leaf and composite share the same interface
	var node FileSystemNode = src
	fmt.Printf("PASS: src dir size via interface = %d\n", node.Size())
}

File: Languages/python/low-level-design-lld/design-patterns/composite.py
"""
Composite Pattern

Composes objects into tree structures. File and Directory share a common
interface, letting clients treat individual objects and compositions uniformly.
"""

from abc import ABC, abstractmethod


class FileSystemNode(ABC):
    def __init__(self, name: str):
        self.name = name

    @abstractmethod
    def size(self) -> int:
        pass

    @abstractmethod
    def display(self, indent: int = 0) -> str:
        pass


class File(FileSystemNode):
    def __init__(self, name: str, size: int):
        super().__init__(name)
        self._size = size

    def size(self) -> int:
        return self._size

    def display(self, indent: int = 0) -> str:
        return f"{'  ' * indent}{self.name} ({self._size}B)"


class Directory(FileSystemNode):
    def __init__(self, name: str):
        super().__init__(name)
        self.children: list[FileSystemNode] = []

    def add(self, node: FileSystemNode) -> "Directory":
        self.children.append(node)
        return self

    def remove(self, name: str) -> None:
        self.children = [c for c in self.children if c.name != name]

    def size(self) -> int:
        return sum(child.size() for child in self.children)

    def display(self, indent: int = 0) -> str:
        lines = [f"{'  ' * indent}{self.name}/"]
        for child in self.children:
            lines.append(child.display(indent + 1))
        return "\n".join(lines)

    def find(self, name: str) -> FileSystemNode | None:
        for child in self.children:
            if child.name == name:
                return child
            if isinstance(child, Directory):
                found = child.find(name)
                if found:
                    return found
        return None


if __name__ == "__main__":
    root = Directory("root")
    src = Directory("src")
    src.add(File("main.py", 1200))
    src.add(File("utils.py", 800))
    root.add(src)
    root.add(File("README.md", 500))

    # Uniform interface: size works on both
    assert root.size() == 2500
    assert src.size() == 2000

    # Display tree
    output = root.display()
    assert "root/" in output
    assert "  src/" in output
    assert "    main.py (1200B)" in output

    # Find across tree
    assert root.find("utils.py") is not None
    assert root.find("nonexistent") is None

    # Remove
    src.remove("utils.py")
    assert src.size() == 1200
    assert root.size() == 1700

    print("All tests passed!")
